### PR TITLE
1941 move EpochStack to TerminationDetector

### LINKED ISSUE #1941: Move epoch stack to TerminationDetector

### PR DESCRIPTION
**Pull request by stmcgovern**
Closes #1941. Add EpockStack to TerminationDetector.

### CODE REVIEW

File: src/vt/termination/termination.impl.h [hunk 1]
```diff
@@ -83,6 +83,104 @@ inline bool TerminationDetector::isDS(EpochType epoch) {
...
+  /*
+   * popEpoch(epoch) shall remove the top entry from epoch_size_, iif the size
+   * is non-zero and the `epoch' passed, if `epoch != no_epoch', is equal to the
+   * top of the `epoch_stack_.top()'; else, it shall remove any entry from the
+   * top of the stack.
+   */
+  auto const& non_zero = epoch_stack_.size() > 0;
+  vtAssertExprInfo(
+    non_zero and (epoch_stack_.top() == epoch.get() or epoch == no_epoch),
+    epoch, non_zero, epoch_stack_.top()
```
Comment (PhilMiller) [inline src/vt/termination/termination.impl.h:172]:
```suggestion
    epoch, non_zero, non_zero ? epoch_stack_.top() : no_epoch
```

File: src/vt/termination/termination.h [hunk 2]
```diff
@@ -71,6 +71,18 @@ namespace vt { namespace term {
...
+struct EpochStack {
+  using DataType = epoch::detail::EpochImplType;
+
+  void push(DataType in) { stack_[cur_++] = in; }
+  DataType top() const { return stack_[cur_-1]; }
+  void pop() { cur_--; }
+  unsigned int size() const { return cur_; }
+
+  int cur_ = 0;
+  std::array<DataType, 64> stack_;
```
Comment (PhilMiller) [inline src/vt/termination/termination.h:83]: We'd discussed making this a potentially dynamically sized structure.

As in the scheduler queue, where we use a circular buffer, we may want to do something clever where we keep the `std::array` for the shallow case, and do a bulk hand-off to a growable structure for the *bottom* of the stack that's not changing as rapidly

Comment (PhilMiller) [reply src/vt/termination/termination.h:83]: I'd suggest opening an issue and doing that in a follow-up commit

**Review by lifflander — APPROVED** [review]
Looks good to me!

File: src/vt/termination/termination.h [hunk 3]
```diff
@@ -102,13 +114,27 @@ struct TerminationDetector :
...
    */
   TerminationDetector();
 
-  virtual ~TerminationDetector() {}
+  virtual ~TerminationDetector() {
+    //Pop all extraneous epochs off the stack greater than 1
+      auto stack_size = epoch_stack_.size();
+      while (stack_size > 1) {
+        stack_size = (epoch_stack_.pop(), epoch_stack_.size());
+       }
```
Comment (PhilMiller) [inline src/vt/termination/termination.h:129]: The way this is written seems overly clever. The variable `stack_size` isn't used outside the loop control, so the whole thing could be written

```suggestion
      // Pop all extraneous epochs off the stack greater than 1
      while (epoch_stack_.size() > 1) {
        epoch_stack_.pop();
      }
```

Also, the indentation was messed up.

That said, this is not a big concern.

File: src/vt/termination/termination.h [hunk 4]
```diff
@@ -772,11 +796,25 @@ struct TerminationDetector :
    */
   static void epochContinueHandler(TermMsg* msg);
 
-private:
+public:
+  inline EpochType getEpoch() const;
+  inline void pushEpoch(EpochType const& epoch);
+  inline EpochType popEpoch(EpochType const& epoch = no_epoch);
```
Comment (PhilMiller) [inline src/vt/termination/termination.h:802]: These should take `epoch` by value

File: src/vt/termination/termination.h [hunk 5]
```diff
@@ -772,11 +796,25 @@ struct TerminationDetector :
...
+
+  inline void pushEpochFast(EpochType epoch) {
+    epoch_stack_.push(epoch.get());
+  }
+  inline void popEpochFast(EpochType epoch) {
```
Comment (PhilMiller) [inline src/vt/termination/termination.h:807]: Why is this version taking the `epoch` argument, if it's not going to check it?

File: src/vt/termination/termination.h (hunk 3 above)
Comment (lifflander) [reply src/vt/termination/termination.h:129]: You can blame the "cleverness" on me (it was copied from AM).

File: src/vt/messaging/active.impl.h [hunk 6]
```diff
@@ -445,54 +445,16 @@ ActiveMessenger::PendingSendType ActiveMessenger::broadcastMsgAuto(
...
-      epoch_stack_.pop(),epoch :
-      no_epoch;
-  }
+  return theTerm()->popEpoch(epoch);
 }
 
 inline EpochType ActiveMessenger::getEpoch() const {
-  return getGlobalEpoch();
+  return theTerm()->getEpoch();
 }
```
Comment (PhilMiller) [inline src/vt/messaging/active.impl.h:459]: I don't really like keeping these wrappers here, but I guess it spares us rewriting all the call sites at the same time. It does also allow for caching `theTerm()` in `ActiveMessenger` for internal usage

**Comment by PhilMiller** [comment]
Merging vs fixing stuff is up to you, @lifflander

File: src/vt/termination/termination.h (hunk 3 above)
Comment (stmcgovern) [reply src/vt/termination/termination.h]: done

File: src/vt/termination/termination.h (hunk 4 above)
Comment (stmcgovern) [reply src/vt/termination/termination.h]: done

File: src/vt/termination/termination.h (hunk 5 above)
Comment (stmcgovern) [reply src/vt/termination/termination.h]: done

File: src/vt/messaging/active.impl.h (hunk 6 above)
Comment (stmcgovern) [reply src/vt/messaging/active.impl.h:459]: I've not changed this. Do you think it's worth an issue to remove them or to go with the caching, @PhilMiller @lifflander ?

Comment (PhilMiller) [reply src/vt/messaging/active.impl.h:459]: Let's leave it be for now, and think about caching if that lookup ever turns up as a hot spot in profiles

**Review by PhilMiller — APPROVED** [review]

**Comment by lifflander** [comment]
Since these problems already existed, I would prefer to merge and then fix them in a separate issue.

**Pull request merged**